lib/services/local/generic/generic_wrapper_test.go

```diff
@@ -0,0 +1,225 @@
+		r.Metadata.Expires = timestamppb.New(cfg.Expires)
+	}
+	return &r, nil
+}
+
+// TestGenericWrapperCRUD tests backend operations with the generic service.
+func TestGenericWrapperCRUD(t *testing.T) {
+	ctx := context.Background()
+
+	memBackend, err := memory.New(memory.Config{
+		Context: ctx,
+		Clock:   clockwork.NewFakeClock(),
+	})
+	require.NoError(t, err)
+
+	const backendPrefix = "generic_prefix"
+
+	service, err := NewServiceWrapper[*testResource153](memBackend,
+		"generic resource",
+		backendPrefix,
+		marshalResource153,
+		unmarshalResource153)
+	require.NoError(t, err)
+
+	// Create a couple test resources.
+	r1 := newTestResource153("r1")
+	r2 := newTestResource153("r2")
+
+	// sanity check: marshal/unmarshal round trips
+	r1data, err := marshalResource153(r1)
+	require.NoError(t, err)
+	r1clone, err := unmarshalResource153(r1data)
+	require.NoError(t, err)
+	require.Equal(t, r1, r1clone)
+
+	// Initially we expect no resources.
+	out, nextToken, err := service.ListResources(ctx, 200, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.Empty(t, out)
+
+	// Create both resources.
+	r1, err = service.CreateResource(ctx, r1)
+	require.NoError(t, err)
+	r2, err = service.CreateResource(ctx, r2)
+	require.NoError(t, err)
+
+	require.NotEmpty(t, r1.GetMetadata().GetRevision())
+	require.NotEmpty(t, r2.GetMetadata().GetRevision())
+	require.NotEqual(t, r1.GetMetadata().GetRevision(), r2.GetMetadata().GetRevision())
+
+	// Fetch all resources using paging default.
+	out, nextToken, err = service.ListResources(ctx, 0, "")
+	require.NoError(t, err)
+	require.Empty(t, nextToken)
+	require.NotEmpty(t, out)
+	require.Equal(t, []*testResource153{r1, r2}, out)
+
+	// Fetch a paginated list of resources.
+	paginatedOut := make([]*testResource153, 0, 2)
+	numPages := 0
+	for {
+		numPages++
+		out, nextToken, err = service.ListResources(ctx, 1, nextToken)
+		require.NoError(t, err)
+
+		paginatedOut = append(paginatedOut, out...)
+		if nextToken == "" {
+			break
+		}
+	}
+
+	require.Equal(t, 2, numPages)
+	require.Equal(t, []*testResource153{r1, r2}, paginatedOut)
+
+	// Fetch a specific service provider.
+	r, err := service.GetResource(ctx, r2.GetMetadata().GetName())
+	require.NoError(t, err)
+	require.Equal(t, r2, r)
+
+	// Try to fetch a resource that doesn't exist.
+	_, err = service.GetResource(ctx, "doesnotexist")
+	require.ErrorIs(t, err, trace.NotFound(`generic resource "doesnotexist" doesn't exist`))
+
+	// Try to create the same resource.
+	_, err = service.CreateResource(ctx, r1)
+	require.ErrorIs(t, err, trace.AlreadyExists(`generic resource "r1" already exists`))
+
+	// Update a resource.
+	r1.Metadata.Labels = map[string]string{"newlabel": "newvalue"}
+	r1, err = service.UpdateResource(ctx, r1)
+	require.NoError(t, err)
+	r, err = service.GetResource(ctx, r1.GetMetadata().GetName())
+	require.NoError(t, err)
+	//nolint:staticcheck // SA1019. Deprecated, but still needed.
+	r.Metadata.Id = r1.Metadata.Id
```

Comment:
Why do we need to copy the Metadata.Id?

Could we use a form of comparison that ignores it instead?

Reply:
I've spent some time wrestling with comparison options, but they all had one issue or another. I'm all ears if you can suggest some working alternative.

Reply:
Yeah, I didn't have much luck using it to compare protobuf messages. It should work in theory, but I couldn't force it to behave and switched to the implementation above. It would be more elegant, so I'm keen to learn if you know the proper way.

Reply:
Ok, I had more luck this time, although I can't say I'm a great fan of having to list every single type in this fashion:
```golang
	ignoreUnexported := cmp.Options{
		cmpopts.IgnoreUnexported(testResource153{}),
		cmpopts.IgnoreUnexported(headerv1.Metadata{}),
		cmpopts.IgnoreUnexported(timestamppb.Timestamp{}),
	}
```

Reply:
There's a proper incantation to compare protos, it's goes like this: https://github.com/gravitational/teleport/blob/31fba7b477452bc957cff73147f9d19946c83222/api/types/device_test.go#L134-L136

There are also protocmp.IgnoreFields, protocmp.SortRepeatedFields (effectively for order-independent compares), and a bunch of other stuff I suppose. There are also plain `cmp` equivalent for non-proto structs.

Reply:
> Ok, I had more luck this time, although I can't say I'm a great fan of having to list every single type in this fashion

Yeah, YMMV.

Reply:
Oh, thanks for pointing out `protocmp.Transform()`, this one will come in handy when comparing plain proto messages. Here it doesn't work on its own, unfortunately, as `testResource153` is _not_ a protobuf message. I could probably rewrite the whole test to use a different resource, or pass some additional options, but I think it is cleaner to use a consistent approach, i.e., I'll leave it at these three options above.